### ISSUE TITLE
Rope crate documentation

### ISSUE THREAD

**Issue by shymega** [google/xi-editor#115]
Hello,

I'm looking into the Rope crate from xi-editor for my similar text editor project that I'm hacking on, and I found that documentation was particularly lacking in a lot of areas.

Perhaps improved documentation could be something we could work towards in regards to Rope? 

Thanks!

**Comment by raphlinus** [comment]
I agree it could use a lot better documentation. A major reason it's not there already is that I don't consider the design fully settled. It's "good enough" but I can see it being refined more.

Keep in mind that `::Rope` is deprecated in favor of `::rope::Rope`. The only reason I haven't deleted the former is that it contains both tests and performance optimizations that aren't fully in place in the latter.

The part of the Rope design I'm least happy about is the role of `Interval`, in particular the role of open and closed endpoints. Take the string `"a\nb"` and a metric counting newlines. In this metric, `[0, 1)` represents `"a\n"`, but I also needed a way to represent the whole substring. I spent a bunch of time struggling with using `[0, 2)` for that (basically representing all intervals as closed-open), but it led to special cases and fudge factors. My hope in adopting open and closed intervals is that `[0, 1]` could represent that case and involve less special-casing in the code. I was also looking ahead to implement [inclusive and exclusive spans](https://developer.android.com/reference/android/text/Spanned.html) analogous to Android, and felt that the same mechanism could encompass that goal as well.

I'm not sure I succeeded. I think there's a decision point here - either revert and use `[0, 2)`, or make the whole thing more mathematically sound. For example, taking the subsequence `(1, 1]` should probably return `"\nb"`, but in the current implementation it doesn't.

I haven't been especially motivated to come back to this, because what I have works well enough for the current xi codebase.

Concepts like "metrics" are tricky and clever. I mean this mostly in the bad sense, that it's far from obvious to someone how to implement a new metric from scratch, especially if it doesn't quite match the assumptions of the existing ones. I'll give a really tough example - let's say that your leaf was a `Vec<u8>` instead of a `String`, with the ability to split a utf-8 sequence across two leaves. Detecting codepoint boundaries fits the "metric" framework (a metric is defined as a monoid homomorphism where the monoid is unsigned integer sum), but the existing code would break, as it basically assumes that boundaries are at the _end_ of a subsequence.

I also want to generalize the metric concept so that it doesn't have to be sum. Boolean "or" is another example of a monoid homomorphism, and is useful for things such as determining whether a substring is all-ASCII. You absolutely can do this by counting non-ASCII characters, but a single bit would be more efficient than a count.

This is a somewhat longwinded way of saying that it's not just a question of writing docs, it deserves a deep re-examination. Also, tests should be done at the same time.

BTW, I have a fair amount of draft writing that explains some concepts and motivation behind xi ropes. I've been meaning to clean them up and publish them (most likely on Medium) but perhaps it would be useful to upload the rough drafts for curious and motivated readers.

**Comment by shymega** [comment]
Hello,

Raph Levien <notifications@github.com> writes:

> I agree it could use a lot better documentation. A major reason it's
> not there already is that I don't consider the design fully
> settled. It's "good enough" but I can see it being refined more.

Yeah, I see what you mean.

> [..snip..]

> This is a somewhat longwinded way of saying that it's not just a
> question of writing docs, it deserves a deep re-examination. Also,
> tests should be done at the same time.

> BTW, I have a fair amount of draft writing that explains some
> concepts and motivation behind xi ropes. I've been meaning to clean
> them up and publish them (most likely on Medium) but perhaps it
> would be useful to upload the rough drafts for curious and motivated
> readers.

That would be great. I look forward to seeing them.

Perhaps you could put the writings into the xi-editor repo as well?
That way people who might not want to visit Medium can still read it
as well.

> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/google/xi-editor/issues/115#issuecomment-238117580

-- 
Regards,
Dom Rodriguez

**Comment by portstrom** [comment]
I'm trying to understand the implementation of ropes, and I find it more difficult than it needs to be. I'm suggesting some ways it can be made easier.

First thing, it would be helpful to have some overview. There's the “rope science” series, which explains the theory behind the rope design, but there doesn't seem to be anything explaining the code. I couldn't even find where to start reading the code, and had to read random code until I found the central parts.

Modules are terribly long. Having them split into smaller modules and trying to keep them as self-contained as possible would be helpful. The most obvious possibility to split modules that I see is to take `Cursor` out of `tree.rs`. `Cursor` takes up half the file and isn't used by any of the other code in the file. The next possibility in `tree.rs` is to take out `Metric`. This would allow comprehending the essential features of the tree without at the same time having to comprehend metrics.

Data structures, constants and functions could be better documented. Currently such a central thing is `NodeBody` for example comes with no documentations at all. Neither do the methods of `Node`. It would be helpful to have some explanation of the entire procedure of modifying a tree, giving some examples and pointing out which methods are involved in each step.

Some things could perhaps be simplified. There is a comment that says “This is here to demonstrate that this is a monoid.”. Demonstrating something doesn't justify the existence of something. There's so much code written to flaunt the author's knowledge of mathematics, design patterns and code fashion, it isn't funny anymore. If the code is there to demonstrate something, the code can be removed, otherwise the comment can be removed. Some methods in `Node` seem to be there only to be used by the `TreeBuilder`. This could could perhaps better be put in the `TreeBuilder` implementation. Or even better, since `TreeBuilder` is only an unnecessary wrapper around `Node`, remove it entirely and move all of its methods into `Node`.

**Comment by cmyr** [comment]
I agree that there's definitely room for improvement here! I've also thought about moving cursor and metric, and adding extensions to tree for the code that relies on them. As always, PRs are welcome.

The documentation situation has improved over time, but it's definitely still not great. There are large parts of the rope (how the cache in the cursor is designed, for instance) that I don't understand myself.
